packages/shadycss/src/shadow-parts.js

```diff
@@ -166,6 +163,134 @@ export function formatShadyPartSelector(
     .join('');
 }
 
+/**
+ * Build a forwarding map from part names in the given host, to part names in
+ * all ancestor scopes that also provide styles for those parts, by walking up
+ * the DOM following "exportparts" attributes on each host element.
+ *
+ * For example, given the <x-c> node from the following tree:
+ *
+ *   #document
+ *     <x-a exportparts="foo">
+ *       #shadow-root
+ *         <x-b exportparts="foo,bar:bar2">
+ *           #shadow-root
+ *             <x-c>
+ *
+ * Then return:
+ *
+ *   {
+ *     foo: [
+ *       ["x-a", "x-b", "foo"],
+ *       ["document", "x-a", "foo"],
+ *     ],
+ *     bar: [
+ *       ["x-a", "x-b", "bar2"],
+ *     ],
+ *  }
+ *
+ * @param {!HTMLElement} host The element whose shadow parts we are evaluating.
+ * @return {!Object<!string, !Array<string>>} An object that maps from part name
+ * in the given `host` to an array of `[providerScope, receiverScope, partName]`
+ * triplets describing additional style sources for that part.
+ *
+ * TODO(aomarks) Cache each host's exportparts mapping so that we don't need to
+ * walk any hosts twice.
+ */
+export function findExportedPartMappings(host) {
+  // The mapping we will return.
+  const result = {};
+
+  // As we walk up the tree, this object always maintains a mapping from part
+  // names in the previously walked scope, to part names in the base host scope.
+  let forwarding;
+
+  let superHost;
+
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const attr = host.getAttribute('exportparts');
+    if (!attr) {
+      break;
+    }
+
+    // Initialize superHost on the first iteration. Done here, inside the loop,
+    // so that we can return as quickly as possible when there is no exportparts
+    // attribute.
+    if (superHost === undefined) {
+      const root = host.getRootNode();
+      if (!root || root === document) {
+        // exportparts doesn't mean anything here; there is nothing beyond the
+        // document scope.
+        break;
+      }
+      superHost = root.host;
+      if (!superHost) {
+        break;
+      }
+    }
+
+    const receiverScope = superHost.localName;
+    const superRoot = superHost.getRootNode();
+    if (!superRoot) {
+      break;
+    }
+
+    let providerScope;
+    let superSuperHost;
+    if (superRoot === document) {
+      providerScope = 'document';
+    } else {
+      superSuperHost = superRoot.host;
+      if (!superSuperHost) {
+        break;
+      }
+      providerScope = superSuperHost.localName;
+    }
+
+    const parsed = parseExportPartsAttribute(attr);
```

Comment:
I was thinking since this is the case where the application has an syntactically invalid part attribute, it should be rare, and so it makes sense to first do checks that are more likely to return.